src/devices.jl

```diff
@@ -0,0 +1,37 @@
+import ClimaComms
+
+# Sometimes, it is convenient to know whether something is running on a device or on its
+# host. We define new AbstractDevices and AbstractCommsContext to identify the case of a
+# function running on the device with data on the device.
+
+"""
+    DeviceSideDevice()
+
+This device represents data defined on the device side of an accelerator.
+
+The most common example is data defined on a GPU. DeviceSideDevice() is used for
+operations within the accelerator.
+"""
+struct DeviceSideDevice <: ClimaComms.AbstractDevice end
+
+
+"""
+    DeviceSideContext()
+
+Context associated to data defined on the device side of an accelerator.
+
+The most common example is data defined on a GPU. DeviceSideContext() is used for
+operations within the accelerator.
+"""
+struct DeviceSideContext <: ClimaComms.AbstractCommsContext end
+
+# Some functions use context.device
```

Comment:
can we just change those?

Reply:
> can we just change those?

What do you mean?

Reply:
https://github.com/CliMA/ClimaCore.jl/blob/a04027449080a29dcaa4e390a56231e6dd3fbded/src/Topologies/topology2d.jl#L127

This is where we define `device(space)` essentially

Reply:
I added the `getproperty` mostly to mimic the interface that `context` has, but if everything is wired correctly, we shouldn't strictly need it. So, I am happy to remove it if you don't like it.

Reply:
Okay, I missed that ClimaComms already defines 
```
device(ctx::SingletonCommsContext) = ctx.device
```